llvm/lib/Frontend/SYCL/OffloadWrapper.cpp

```diff
@@ -0,0 +1,527 @@
+    SmallVector<Constant *> WrappedImages;
+    WrappedImages.reserve(Images.size());
+    for (size_t I = 0, E = Images.size(); I != E; ++I)
+      WrappedImages.push_back(wrapImage(Images[I], Twine(I), OffloadKindTag));
+
+    return combineWrappedImages(WrappedImages, OffloadKindTag);
+  }
+
+  void createRegisterFatbinFunction(GlobalVariable *FatbinDesc) {
+    auto *FuncTy = FunctionType::get(Type::getVoidTy(C), /*isVarArg*/ false);
+    auto *Func = Function::Create(FuncTy, GlobalValue::InternalLinkage,
+                                  Twine("sycl") + ".descriptor_reg", &M);
+    Func->setSection(".text.startup");
+
+    // Get RegFuncName function declaration.
+    auto *RegFuncTy =
+        FunctionType::get(Type::getVoidTy(C), PointerType::getUnqual(C),
+                          /*isVarArg=*/false);
+    FunctionCallee RegFuncC =
+        M.getOrInsertFunction("__sycl_register_lib", RegFuncTy);
+
+    // Construct function body
+    IRBuilder Builder(BasicBlock::Create(C, "entry", Func));
+    Builder.CreateCall(RegFuncC, FatbinDesc);
+    Builder.CreateRetVoid();
+
+    // Add this function to constructors.
+    appendToGlobalCtors(M, Func, /*Priority*/ 1);
+  }
+
+  void createUnregisterFunction(GlobalVariable *FatbinDesc) {
+    auto *FuncTy = FunctionType::get(Type::getVoidTy(C), /*isVarArg*/ false);
+    auto *Func = Function::Create(FuncTy, GlobalValue::InternalLinkage,
+                                  "sycl.descriptor_unreg", &M);
+    Func->setSection(".text.startup");
+
+    // Get UnregFuncName function declaration.
+    auto *UnRegFuncTy =
+        FunctionType::get(Type::getVoidTy(C), PointerType::getUnqual(C),
+                          /*isVarArg=*/false);
+    FunctionCallee UnRegFuncC =
+        M.getOrInsertFunction("__sycl_unregister_lib", UnRegFuncTy);
+
+    // Construct function body
+    IRBuilder<> Builder(BasicBlock::Create(C, "entry", Func));
+    Builder.CreateCall(UnRegFuncC, FatbinDesc);
+    Builder.CreateRetVoid();
+
+    // Add this function to global destructors.
+    appendToGlobalDtors(M, Func, /*Priority*/ 1);
+  }
+}; // end of Wrapper
+
+} // anonymous namespace
+
+Error llvm::offloading::sycl::wrapSYCLBinaries(llvm::Module &M,
+                                               ArrayRef<ArrayRef<char>> Buffers,
+                                               SYCLWrappingOptions Options) {
+  Wrapper W(M, Options);
+  SmallVector<std::unique_ptr<OffloadBinary>> OffloadBinaries;
+  OffloadBinaries.reserve(Buffers.size());
+  SmallVector<OffloadingImage> Images;
+  Images.reserve(Buffers.size());
+  for (auto Buf : Buffers) {
+    MemoryBufferRef MBR(StringRef(Buf.begin(), Buf.size()), /*Identifier*/ "");
+    auto OffloadBinaryOrErr = OffloadBinary::create(MBR);
+    if (!OffloadBinaryOrErr)
+      return OffloadBinaryOrErr.takeError();
+
+    OffloadBinaries.emplace_back(std::move(*OffloadBinaryOrErr));
+    Images.emplace_back(OffloadBinaries.back()->getOffloadingImage());
```

Comment:
@asudarsa Does there need to be one more layer of unwrapping, or am I misunderstanding how the process works? My understanding is that clang-sycl-linker outputs an a list of OffloadingImages concatenated together. Then, clang-linker-wrapper takes the output from the device linker and then puts it in an OffloadingImage. So in the SYCL case, this functinon receives a list of OffloadingImages, with each of these OffloadingImages holding a list of OffloadingImages concatenated together?